### ISSUE TITLE
redis output support

### ISSUE THREAD

**Issue by vkhatri** [elastic/filebeat#132]
Hello! Thank you for releasing `1.0.0-beat4`! 

I noticed that `redis output` has been removed from the default configuration from other beats too and wanted to know whether `redis output` will be added in the future.

It was a major blocker for us to use `LSF`, having `redis output` will really help us in streamlining our E{L,P,T}K setup. I tried to look up for a road map, but could not find any info on it. 

Thanks again!

**Comment by tsg** [comment]
The Redis output is currently deprecated but you can still use it for the time being. Just be aware that we don't have too many automated tests for it at the moment.

To share a bit of our plans and why we deprecated the Redis output, the way we're suggesting people to add a queue to the system at the moment is following something like this:

    many Beats ---> Logstash ---> queue (Redis, Kafka, etc.) ---> Logstash ---> Elasticsearch

The Logstash instance (or group of instances) that sits in front of the queue doesn't have to do any processing so it will typically be quite fast and be able to handle the traffic of many Beats. After this gateway-Logstash, you can use many of the output plugins that Logstash support to use any Queue system that we have. Eventually Logstash will have its own persistent queue as well.

Let me share also why we prefer not having a Beats -> Redis,Kafka,RabbitMQ,etc. connection:

While it sounds simple at the begging to "just add another output", the maintenance costs are really quite high. You need to setup integration tests which tend to be slow, keep up to date with the version of the downstream, understand the error conditions of each queue and do the right thing, support different load balancing techniques. Maybe today it looks like supporting 3-4 queuing systems would be enough for 90% of the use cases, but the landscape keeps changing and before we know it we might have much more to support.

This cost is payed already in Logstash which supports 200 plugins and trust us that's no easy task.

Instead we plan to put quite a lot of thought and effort in the communication between the Beats and Logstash, to make sure it supports encryption, client authentication, sane back pressure, sane error handling, sane load balancing. We think that having a trustworthy link between the forwarders and the queue system justifies the cost of installing an extra Logstash instance.

I'd really like to hear your (and others) thoughts on the above, as nothing is set is stone. I'd be in particular interested to know if you have any concerns about deploying Logstash in front of the queuing system, perhaps we can work on addressing them.

**Comment by vkhatri** [comment]
@tsg I will check out Logstash persistent queue, sound like a better solution. The only reason we were planning to use Redis is to have a broker cache in case logstash consumer become unavailable.

Really appreciate your detailed response and it is good to know which direction `beats` are going altogether. I am pretty sure that at some point we would have looked at replacing Redis with Kafka. But with front end logstash + persistent queue, setup looks simple and easy to scale with consumer logstash output.

@rudijs thank you for sharing your setup details, we can definitely leverage containers at some places.

thanks!

**Comment by tmatei** [comment]
@tsg While I understand the rationale behind this, I don't agree with this decision. I was looking forward to use filebeat as a lightweight forwarder, but without the Logstash persistence this is a bad idea. The whole idea of having a queue is to be able to buffer messages in case an end point is down. This makes an assumption that logstash will always be up and that is not always the case. I would strongly suggest keeping support for some sort of queue system, redis or rabbitmq (preferrably both).

**Comment by congpine** [comment]
I have setup the systems as @tsg suggested above and it seems to be working fine. I haven't got many servers to send logs yet, so i'm not sure how it will perform. 
The things that worry me most is when Logstash Indexer couldn't get input or output to Elasticsearch, it will crash and may be require manually fix, restart. we may lose logs in Redis if it is over-filled. 
Will Kafka or RabbitMQ offer writing to disk after it hits a certain time or queue size?

**Comment by tmatei** [comment]
any thoughts about possibly supporting a queue/buffer mechanism in between?

**Comment by CliveJL** [comment]
I really like the idea of continuing to support a Redis output in the lightweight Beats agent. We have a Production environment processing a pipeline of tens of thousands of logs per second, and the solution that works well for us is to get the logs off of the application servers quickly with low overhead, and into large Redis queues, and then have them processed by an auto-scaling Pool of Logstash Consumers before passing them into Elasticsearch. Having to manage another layer of Logstash instances between the Beats and Redis layers seems like unnecessary overhead, with more moving parts.

In some areas we use the Beaver agent as it is lighter than Logstash, and still supports output to Redis. I feel it would be a step in the wrong direction for Beats if it were to lose the ability to output to a broker like Redis. Just my humble upvote...

**Comment by johntdyer** [comment]
+1 for keeping support for some sort of queue.

**Comment by tfendt** [comment]
Vote for keeping support. I was all excited to remove the Logstash shipper with Filebeat and found this page after reading it was depreciated :(

Less servers in the mix means less things to potentially go wrong.

**Comment by devopsprosiva** [comment]
+1 to have beats -> redis support without a logstash instance in between.

**Comment by Zurkdahool** [comment]
+1 (to have beats -> redis support)

**Comment by mikev-mogo** [comment]
+1 (beats -> redis)

**Comment by broxio** [comment]
+1 to have beats -> redis

**Comment by Krast76** [comment]
+1 to have beats output on redis !

**Comment by bensur** [comment]
@tsg I have a very large system(s) and wanted to migrate into using file beats, the fact that it don't have the Redis support forced me to re-consider the whole thing.
I'm looking to have a simple lightweight sender on the server that doe's not consume too much memory/cpu (didn't fully tested filebeat yet in that aspect) but I can't install logstash on the server as well, I also can't install a logstash instance that will handle all the requests sent from the filebeats (too many data & single point of failure).
So basically I have to go with another solution to gather all my logs :-1:

**Comment by tsg** [comment]
@bensur Typically when gathering logs you also need to parse them, and for parsing you need Logstash anyway. Because Filebeat does synchronous sending, a simple Filebeat -> Logstash -> Elasticsearch setup is all you need for most cases, you don't really need a dedicated queue. If there is back-pressure from Elasticsearch or Logstash, Filebeat simply waits before advancing into files. Essentially, the files themselves are the queue.

In addition, when using Redis, we cannot guarantee that messages are not lost in case of Redis restarts, crashes or network partitions. We can guarantee this in the Filebeat -> Logstash -> Elasticsearch setup.

Btw, you can deploy multiple Logstash servers and Filebeat can do failover and load balancing between them, so it's not a SPOF.

**Comment by Zurkdahool** [comment]
@tsg how does this work in packetbeat  and/or topbeat where do i have no files?

**Comment by tsg** [comment]
@Zurkdahool packetbeat and topbeat have an internal in-memory buffer, but if this gets filled the messages are dropped. However, you can use Packetbeat or Topbeat's file output and Filebeat on the same machine if you want "at least once" delivery. We do have plans for libbeat to support spooling to disk to simplify this.

**Comment by tmatei** [comment]
@tsg I can't quite agree with you on that. We have an issue where we currently log into syslog and if we let it write to disk it starts bringing the load up on the servers due to too many writes. So having redis as a buffer is awesome. This is especially true for packetbeat, since our servers will handle more packets than it can write to disk (in case of downtime). I don't see why you would need to worry about redis downtime. I would say redis is more reliable than logstash at this point. I hope you change your mind and keep redis support.

**Comment by pemontto** [comment]
:+1:  for Redis support

**Comment by jsosic** [comment]
:+1: for Redis

**Comment by ruflin** [comment]
Thanks for all the inputs and discussions. Based on this and some additional discussions we decided to undeprecate Redis output (https://github.com/elastic/beats/pull/1008) and add additional outputs (https://github.com/elastic/beats/issues/943). Closing this issue.

**Comment by svilcu** [comment]
beats->rabbitMQ would be nice.  
We actually have logs shipped via TCP to syslog-ng->rabbitmq-> logstash->elastic
